src/Common/TournamentTree.h

```diff
@@ -0,0 +1,198 @@
+#pragma once
+
+#include <base/defines.h>
+
+#include <utility>
+#include <vector>
+
+#include <Common/BitHelpers.h>
+
+
+namespace DB
+{
+
+/**
+  *
+  */
+template <typename Cursor, typename Compare = std::less<Cursor>>
+class TournamentTree
+{
+public:
+    TournamentTree() = default;
+
+    explicit TournamentTree(const std::vector<Cursor> & cursors_, Compare compare_ = Compare())
+        : compare(compare_)
+    {
+        for (auto & cursor : cursors_)
+            cursors.emplace_back(cursor);
+
+        if (cursors.empty())
+        {
+            tournament_tree.resize(2);
+            return;
+        }
+
+        buildTree();
+    }
+
+    explicit TournamentTree(std::vector<Cursor> && cursors_, Compare compare_ = Compare())
+        : compare(compare_)
+    {
+        for (auto & cursor : cursors_)
+            cursors.emplace_back(std::move(cursor));
+
+        if (cursors.empty())
+        {
+            tournament_tree.resize(2);
+            return;
+        }
+
+        buildTree();
+    }
+
+    ALWAYS_INLINE bool isValid() const { return tournament_tree[1] != InvalidTreeNodeValue; }
+
+    ALWAYS_INLINE Cursor & current()
+    {
+        assert(isValid());
+
+        size_t cursor_index = cursorIndexFromTreeNodeValue(tournament_tree[1]);
+        return cursors[cursor_index];
+    }
+
+    ALWAYS_INLINE void next()
+    {
+        assert(isValid());
+
+        size_t cursor_index = cursorIndexFromTreeNodeValue(tournament_tree[1]);
+        cursors[cursor_index].next();
+        replayGames(cursor_index);
+    }
+
+    ALWAYS_INLINE Cursor & getCursor(size_t cursor_index) { return cursors[cursor_index]; }
+
+    ALWAYS_INLINE void updateCursor(size_t cursor_index) { replayGames(cursor_index); }
+
+private:
+    ALWAYS_INLINE void buildTree()
```

Comment:
This is pretty aggressive inlining.

Apart from affecting build time, did you measure the difference in performance with and without explicit inlining?
Too much inlining can negatively affect the performance so I'm just interested if you tested it out.